criu/sigact.c

```diff
@@ -0,0 +1,319 @@
+#include "types.h"
+#include "infect.h"
+#include "protobuf.h"
+#include "pstree.h"
+#include "parasite.h"
+#include "restorer.h"
+
+/*
+ * If parent's sigaction has blocked SIGKILL (which is non-sense),
+ * this parent action is non-valid and shouldn't be inherited.
+ * Used to mark parent_act* no more valid.
+ */
+static rt_sigaction_t parent_act[SIGMAX];
+#ifdef CONFIG_COMPAT
+static rt_sigaction_t_compat parent_act_compat[SIGMAX];
+#endif
+
+static bool sa_inherited(int sig, rt_sigaction_t *sa)
+{
+	rt_sigaction_t *pa;
+	int i;
+
+	if (current == root_item)
+		return false; /* XXX -- inherit from CRIU? */
+
+	pa = &parent_act[sig];
+
+	/* Omitting non-valid sigaction */
+	if (pa->rt_sa_mask.sig[0] & (1 << SIGKILL))
+		return false;
+
+	for (i = 0; i < _KNSIG_WORDS; i++)
+		if (pa->rt_sa_mask.sig[i] != sa->rt_sa_mask.sig[i])
+			return false;
+
+	return pa->rt_sa_handler == sa->rt_sa_handler && pa->rt_sa_flags == sa->rt_sa_flags &&
+	       pa->rt_sa_restorer == sa->rt_sa_restorer;
+}
+
+static void *stack32;
+
+#ifdef CONFIG_COMPAT
+static bool sa_compat_inherited(int sig, rt_sigaction_t_compat *sa)
+{
+	rt_sigaction_t_compat *pa;
+	int i;
+
+	if (current == root_item)
+		return false;
+
+	pa = &parent_act_compat[sig];
+
+	/* Omitting non-valid sigaction */
+	if (pa->rt_sa_mask.sig[0] & (1 << SIGKILL))
+		return false;
+
+	for (i = 0; i < _KNSIG_WORDS; i++)
+		if (pa->rt_sa_mask.sig[i] != sa->rt_sa_mask.sig[i])
+			return false;
+
+	return pa->rt_sa_handler == sa->rt_sa_handler && pa->rt_sa_flags == sa->rt_sa_flags &&
+	       pa->rt_sa_restorer == sa->rt_sa_restorer;
+}
+
+static int restore_compat_sigaction(int sig, SaEntry *e)
+{	
+	extern rt_sigaction_t sigchld_act;
```

Comment:
should be in the header file

Reply:
This is a work around for aarch64 builds. For whatever reason, defining this as a global extern variable doesn't compile on them.

Reply:
I think you messed up something. Could reproduce the problem and post a compile error here?

Reply:
I don't have a aarch64 device to test on, however I do have a [log](https://cirrus-ci.com/task/5783245267140608) of cirrus-ci about this for my previous commit [[a3e7da7](https://github.com/checkpoint-restore/criu/commit/a3e7da7caf05aaf45d0b8f14068ddddbeaa32a50)](https://github.com/checkpoint-restore/criu/commit/a3e7da7caf05aaf45d0b8f14068ddddbeaa32a50)

```
criu/sigact.c: In function 'restore_native_sigaction':
criu/sigact.c:134:17: error: 'sigchld_act' undeclared (first use in this function)
  134 |                 sigchld_act = act;
      |                 ^~~~~~~~~~~
criu/sigact.c:134:17: note: each undeclared identifier is reported only once for each function it appears in
criu/sigact.c: At top level:
cc1: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option '-Wno-dangling-pointer' may have been intended to silence earlier diagnostics
make[3]: *** [/tmp/cirrus-ci-build/scripts/nmk/scripts/build.mk:119: criu/sigact.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [criu/Makefile:76: criu/built-in.o] Error 2
make[1]: *** [Makefile:277: criu] Error 2
make[1]: Leaving directory '/tmp/cirrus-ci-build'
real	0m10.878s
user	0m29.333s
sys	0m7.986s
make: *** [Makefile:2: local] Error 2
make: Leaving directory '/tmp/cirrus-ci-build/scripts/ci'
Exit status: 2
```

You can clearly see that it was defined: https://github.com/checkpoint-restore/criu/commit/a3e7da7caf05aaf45d0b8f14068ddddbeaa32a50#diff-fd7334c98ae8f16e45b62b4221d1c69b5b7d948bf06e8ad6c42863a4de092107R69

Reply:
Nevermind, I messed up (basically I defined it within a ifdef).
Disregard my comment.